### PR TITLE
migrating word embeddings [WIP]

### PR DESCRIPTION
**Pull request by ibrahimsharaf**
- Issue (https://github.com/marco-c/crashsimilarity/issues/30)
- Created an abstract base class for word embeddings for easier extension with new word embedding algorithms, and 2 subclasses for word2vec and doc2vec.
- Migrated word vectors code from crash_similarity.py to word_embeddings.py

### CODE REVIEW

**Comment by ibrahimsharaf** [comment]
@marco-c what do you think?, should I continue and fix the crash similarity unit tests? (it has crashed because I moved the functions as read_corpus, train_model to the new embeddings file).

**Comment by KirillTim** [comment]
I'm not a maintainer, but I have some thoughts about this PR.
It will be very nice if you will add tests and fix from the Issue: https://github.com/marco-c/crashsimilarity/issues/61
Apart from tons of small code quality flaws(methods/variables naming, prints instead of logger calls, ...), IMO it is better to extract corpus reading and, maybe, training to special reader class. Python supports multiple inheritance and this separation will help others to implement other(not based on word2vec) models

**Comment by ibrahimsharaf** [comment]
Thank you @KirillTim for reviewing,
- I didn't look into (https://github.com/marco-c/crashsimilarity/issues/61) yet, but I think it can be fixed in a separate pull request once this one is merged.
- I can't see how the process of training, reading corpus can be generalized, as every embedding algorithm prepares the corpus in its way, e.g: doc2vec required creating a TaggedDocument object, so what can be really generalized in my opinion is only the few lines which read the corpus, same as training models, as each model gets trained in a distinct way using distinct functions.
- will fix the logging and naming thing.

**Comment by KirillTim** [comment]
>>so what can be really generalized in my opinion is only the few lines which read the corpus
Yeah, it will be good. Not only it reduce the code lines but it will also help us if we want to change reading some day in the future

**Comment by KirillTim** [comment]
> I didn't look into (#61) yet, but I think it can be fixed in a separate pull request once this one is merged.

On the one hand, yes it can be fixed in another PR but, on the other hand, if you make PR to master you code should be well tested or at least should works correct.

**Comment by ibrahimsharaf** [comment]
> Yeah, it will be good. Not only it reduce the code lines but it will also help us if we want to change reading some day in the future 

Yes, I agree, will work on it.

> You can make abstract class Algorithm(or whatever name you want) with method train in it. Not every algorithm should inherit from EmbeddingAlgo. It is better to have more abstract class with very basic method train 

I am not sure I understand the benefit out of this, if you mean the word2vec, doc2vec train_model methods by "the very basic method train", we will need to extend it too, because there are differences between the 2 methods.

> On the one hand, yes it can be fixed in another PR but, on the other hand, if you make PR to master you code should be well tested or at least should works correct.

I think I'll wait for @marco-c on this one, as I don't want to populate the PR with other issues to review.

**Comment by KirillTim** [comment]
> I am not sure I understand the benefit out of this, if you mean the word2vec, doc2vec train_model methods by "the very basic method train", we will need to extend it too, because there are differences between the 2 methods.

I mean we can have another algorithms, based on, for example plane strings search [small playground](https://github.com/KirillTim/crashsimilarity/blob/stupid-string-operations/StupidStrings.ipynb). While it is an algorithm it should extend Algorithm class and should also have method `train`, which can be empty, of course

File: word_embeddings.py [hunk 1]
```diff
@@ -0,0 +1,318 @@
...
+                if words1 == words2 or frozenset([tuple(words1), tuple(words2)]) in already_processed:
+                    continue
+                already_processed.add(frozenset([tuple(words1), tuple(words2)]))
+
+                similarities.append((doc1, doc2, self.wmdistance(model, words1, words2, distances)))
+
+        return sorted(similarities, key=lambda v: v[2])
+
+
+class Word2Vec(EmbeddingAlgo):
```
Comment (marco-c) [inline word_embeddings.py]: There's a lot of duplicated code between the `Word2Vec` and `Doc2Vec` classes. It should be possible to put some common things in utility functions or in the base class.
E.g. there could be a `read_traces` in the base class (or as a utility function) that does something like this:
```Python
def read_traces():
    already_selected = set()
    for line in utils.read_files(self.fnames):
        data = json.loads(line)
        proto_signature = data['proto_signature']

        if crash_similarity.should_skip(proto_signature):
            continue

        processed = crash_similarity.preprocess(proto_signature)

        if frozenset(processed) not in already_selected:
            yield elems
```

Then you can use it in `Word2Vec` and `Doc2Vec`, e.g. for `Word2Vec`:
```Python
def read_corpus():
    return [elem for elem in self.read_traces()]
```

**Comment by marco-c** [comment]
Fixing #61 should be in a separate PR, but it should be done and merged before this PR so that we can test that it works as expected.

**Comment by ibrahimsharaf** [comment]
@marco-c I've refactored the code, will push it after merging wmdistance fixes.

**Comment by marco-c** [comment]
> @marco-c I did some refactoring, will push code after merging wmdistance fixes.

Cool, thanks!

**Pull request closed without merging**